client/internal/peer/conn_monitor.go

```diff
@@ -0,0 +1,181 @@
+	ticker := time.NewTicker(signalerMonitorPeriod)
+	defer ticker.Stop()
+
+	lastReady := true
+	for {
+		select {
+		case <-ticker.C:
+			currentReady := conn.signaler.Ready()
+			if !lastReady && currentReady {
+				select {
+				case signalerReady <- struct{}{}:
+				default:
+				}
+			}
+			lastReady = currentReady
+		case <-conn.ctx.Done():
+			return
+		}
+	}
+}
+
+// monitorLocalCandidatesChanged monitors the local candidates and triggers reconnect when they change
+func (conn *Conn) monitorLocalCandidatesChanged(localCandidatesChanged chan<- struct{}) {
+	// TODO: make this global and not per-conn
+
+	ufrag, pwd, err := generateICECredentials()
+	if err != nil {
+		conn.log.Warnf("Failed to generate ICE credentials: %v", err)
+		return
+	}
+
+	ticker := time.NewTicker(candidatesMonitorPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := conn.handleCandidateTick(localCandidatesChanged, ufrag, pwd); err != nil {
+				conn.log.Warnf("Failed to handle candidate tick: %v", err)
+			}
+		case <-conn.ctx.Done():
+			return
+		}
+	}
+}
+
+func (conn *Conn) handleCandidateTick(localCandidatesChanged chan<- struct{}, ufrag string, pwd string) error {
+	conn.log.Debugf("Gathering ICE candidates")
+
+	transportNet, err := newStdNet(conn.iFaceDiscover, conn.config.ICEConfig.InterfaceBlackList)
+	if err != nil {
+		conn.log.Errorf("failed to create pion's stdnet: %s", err)
+	}
+
+	agent, err := newAgent(conn.config, transportNet, candidateTypes(), ufrag, pwd)
+	if err != nil {
+		return fmt.Errorf("create ICE agent: %w", err)
+	}
+	defer func() {
+		if err := agent.Close(); err != nil {
+			conn.log.Warnf("Failed to close ICE agent: %v", err)
+		}
+	}()
+
+	gatherDone := make(chan struct{})
+	err = agent.OnCandidate(func(c ice.Candidate) {
+		log.Debugf("Got candidate: %v", c)
+		if c == nil {
+			close(gatherDone)
+		}
+	})
+	if err != nil {
+		return fmt.Errorf("set ICE candidate handler: %w", err)
+	}
+
+	if err := agent.GatherCandidates(); err != nil {
+		return fmt.Errorf("gather ICE candidates: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(conn.ctx, candidatedGatheringTimeout)
+	defer cancel()
+
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("wait for gathering: %w", ctx.Err())
+	case <-gatherDone:
+	}
+
+	candidates, err := agent.GetLocalCandidates()
+	if err != nil {
+		return fmt.Errorf("get local candidates: %w", err)
+	}
+	log.Debugf("Got candidates: %v", candidates)
+
+	if changed := conn.updateCandidates(candidates); changed {
+		select {
+		case localCandidatesChanged <- struct{}{}:
+		default:
+		}
+	}
+
+	return nil
+}
+
+func (conn *Conn) updateCandidates(newCandidates []ice.Candidate) bool {
+	conn.candidatesMu.Lock()
+	defer conn.candidatesMu.Unlock()
+
+	if len(conn.currentCandidates) != len(newCandidates) {
+		conn.currentCandidates = newCandidates
+		return true
+	}
+
+	for i, candidate := range conn.currentCandidates {
+		if candidate.String() != newCandidates[i].String() {
```

Comment:
We should consider a few things:
1. relay candidates will always be different
2. srflx candidates might have different ports if behind hard NAT

Let's add skip relay candidates and handle srflx better?

Reply:
>1. relay candidates will always be different
good point. I'll remove it

>2. srflx candidates might have different ports if behind hard NAT
I was contemplating using only host candidates but thought a different srflx could open up the possibility for a P22 connection. For some clients, this will result in infinite retries as well. Any suggestion what to do with srflx?

Reply:
I think that just checking for `candidate.Address()` is enough.

Reply:
using `candidateTypesP2P()` might be better to avoid Relay

Reply:
done